include/aspect/particle/property/cpo.h

```diff
@@ -0,0 +1,362 @@
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file LICENSE.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _aspect_particle_property_cpo_h
+#define _aspect_particle_property_cpo_h
+
+#include <aspect/particle/property/interface.h>
+#include <aspect/simulator_access.h>
+#include <array>
+
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
+#include <boost/random.hpp>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Property
+    {
+      enum class DeformationType
+      {
+        Passive
+      };
+      enum class DeformationTypeSelector
+      {
+        Passive
+      };
+
+      enum class AdvectionMethod
+      {
+        ForwardEuler, BackwardEuler, CrankNicolson
+      };
+
+      enum class CPODerivativeAlgorithm
+      {
+        SpinTensor, DRex2004
+      };
+
+      /**
+       * The plugin manages and computes the evolution of Lattice/Crystal Preffered Orientations (LPO/CPO)
+       * on particles. Each ASPECT particle can contain many grains. Each grains contain a size and a orientation
+       * matrix. This allows for LPO evolution tracking with Polycrysal kinematic CPO evolution models such
+       * as D-Rex (Kaminski and Ribe, 2001; ́Kaminski et al., 2004).
+       *
+       * The layout of the data vector per perticle is the following (note that for this plugin the following dim's are always 3):
+       * 1. M mineral times
+       *    1.1  Mineral deformation type   -> 1 double, at location
+       *                                      => cpo_data_position + 0 + mineral_i * (n_grains * 10 + 2)
+       *    2.1. Mineral volume fraction    -> 1 double, at location
+       *                                      => cpo_data_position + 1 + mineral_i * (n_grains * 10 + 2)
+       *    2.2. N grains times:
+       *         2.1. volume fraction grain -> 1 double, at location:
+       *                                      => cpo_data_position + 2 + grain_i * 10 + mineral_i * (n_grains * 10 + 2)
+       *         2.2. a_cosine_matrix grain -> 9 (Tensor<2,dim>::n_independent_components) doubles, starts at:
+       *                                      => cpo_data_position + 3 + grain_i * 10 + mineral_i * (n_grains * 10 + 2)
+       *
+       * Last used data entry is cpo_data_position  + n_minerals * (n_grains * 10 + 2). In case the Crank-Nickelson
+       * advection scheme is used, the previous deriative data is stored behind this. It stores per grain one
+       * double for the volume fraction derivative and 9 doubles for the rotation matrix derivative. This results
+       * the last data entry in this case being:
+       * cpo_data_position + n_minerals * (n_grains * 10 + 2) + mineral_n * (n_grains * 10).
+       *
+       * An other note is that we store exactly the same amount of all minerals (e.g. olivine and enstatite
```

Comment:
It is just a rotation matrix (it is a direction cosine matrix, and the letter used was A in the equation I think, so it became a_cosine_matrix). I will change it to rotation matrix, since I think that is the most clear. .

Reply:
Thanks :). Again, I don't mind the name particularly, but some in line comment would be nice to show how the matrix relates to rotation of the crystal lattice.